hermes/bin/src/logger.rs

```diff
@@ -0,0 +1,101 @@
+//! Setup for logging for the service.
+
+use std::str::FromStr;
+
+use clap::ValueEnum;
+extern crate derive_more;
+use derive_more::Display;
+use tracing::{level_filters::LevelFilter, subscriber::SetGlobalDefaultError};
+use tracing_subscriber::{
+    fmt::{format::FmtSpan, time},
+    FmtSubscriber,
+};
+
+use crate::runtime_extensions::bindings::hermes::logging::api::Level;
+
+/// All valid logging levels.
+#[derive(ValueEnum, Clone, Copy, Display)]
+pub(crate) enum LogLevel {
+    /// Errors
+    #[display(fmt = "Error")]
+    Error,
+    /// Warnings
+    #[display(fmt = "Warn")]
+    Warn,
+    /// Informational Messages
+    #[display(fmt = "Info")]
+    Info,
+    /// Debug messages
+    #[display(fmt = "Debug")]
+    Debug,
+    /// Tracing
+    #[display(fmt = "Trace")]
+    Trace,
+}
+
+impl FromStr for LogLevel {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Error and Warn levels are force to Info level
+        // as Info is the highest log level one can choose.
+        match s {
+            "error" | "warn" | "info" => Ok(LogLevel::Info),
+            "debug" => Ok(LogLevel::Debug),
+            "trace" => Ok(LogLevel::Trace),
+            _ => Err(()),
+        }
+    }
+}
+
+impl From<Level> for LogLevel {
+    fn from(level: Level) -> Self {
+        // Error and Warn levels are force to Info level
+        // as Info is the highest log level one can choose.
+        match level {
+            Level::Info | Level::Warn | Level::Error => LogLevel::Info,
+            Level::Debug => LogLevel::Debug,
+            Level::Trace => LogLevel::Trace,
+        }
+    }
+}
+
+/// Implements a conversion from `LogLevel` enum to the `tracing::Level`.
+impl From<LogLevel> for tracing::Level {
+    fn from(val: LogLevel) -> Self {
+        match val {
+            LogLevel::Error => Self::ERROR,
+            LogLevel::Warn => Self::WARN,
+            LogLevel::Info => Self::INFO,
+            LogLevel::Debug => Self::DEBUG,
+            LogLevel::Trace => Self::TRACE,
+        }
+    }
+}
+
+/// Initializes the subscriber for the logger with the following features.
+/// - JSON format
+/// - Display event level
+/// - Display thread names and ids
+/// - Display event's source code file path and line number
+/// - Display time in RFC 3339 format
+/// - Events emit when the span close
+/// - Maximum verbosity level
+#[allow(dead_code)]
+pub(crate) fn init(
+    log_level: LogLevel, with_thread: bool, with_file: bool, with_line_num: bool,
```

Comment:
Would it be a good idea to encapsulate properties within a separate struct? 
For example:

```rust
pub struct LoggerConfig {
    log_level: LogLevel,
    with_thread: bool,
    with_file: bool,
    with_line_num: bool,
}
```
We also can implement defaults and make fields optional:

```rust
impl LoggerConfig {
    // Entry point to start building
    pub fn builder() -> LoggerConfigBuilder {
        LoggerConfigBuilder::default()
    }
}

#[derive(Default)]
pub struct LoggerConfigBuilder {
    log_level: Option<LogLevel>,
    with_thread: Option<bool>,
    with_file: Option<bool>,
    with_line_num: Option<bool>,
}

impl LoggerConfigBuilder {
    pub fn log_level(mut self, level: LogLevel) -> Self {
        self.log_level = Some(level);
        self
    }
    
    // rest of implementation 
    .
    .
    .
      pub fn build(self) -> LoggerConfig {
        LoggerConfig {
            log_level: self.log_level.unwrap_or(LogLevel::Info), // Provide default value
            with_thread: self.with_thread.unwrap_or(true),
            with_file: self.with_file.unwrap_or(true),
            with_line_num: self.with_line_num.unwrap_or(true),
        }
    }
}
```

then we can initialize configs like this:

```rust
let config = LoggerConfig::builder()
    .log_level(LogLevel::Warn)
    .with_file(false)
    .build();
```
OR

```rust
let config = LoggerConfig::default();
```
OR

```rust
let config = LoggerConfig {
    log_level: LogLevel::Warn,
    ..LoggerConfig::default()
};
```
an call a function like this:

```rust
init(config)?;
```

Reply:
Did some cleanup